### ISSUE TITLE
Cannot connect to redis server with a strong password

### ISSUE THREAD

**Issue by jean-christophe-manciot** [ntop/ntopng#6832]
## Describe the bug
Ntopng cannot connect to redis server and wrongly reports 'ERROR: NOAUTH Authentication required' although it is required and successfully tested with redis-cli.

## Environment
* OS name:Ubuntu
* OS version: 22.04
* Architecture: amd64
* ntopng version/revision: 5.2.220519

**What happened**:
redis
```
systemctl status redis-server
● redis-server.service - Advanced key-value store
     Loaded: loaded (/lib/systemd/system/redis-server.service; enabled; vendor preset: enabled)
     Active: active (running) since Mon 2022-08-29 14:33:45 UTC; 1s ago
       Docs: http://redis.io/documentation,
             man:redis-server(1)
   Main PID: 2195287 (redis-server)
     Status: "Ready to accept connections"
      Tasks: 5 (limit: 9398)
     Memory: 2.8M
        CPU: 42ms
     CGroup: /system.slice/redis-server.service
             └─2195287 "/usr/bin/redis-server 127.0.0.1:6379" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "">

Aug 29 14:33:45 lb-kvm systemd[1]: Starting Advanced key-value store...
Aug 29 14:33:45 lb-kvm systemd[1]: Started Advanced key-value store.
```
ntopng
```
systemctl status ntopng
× ntopng.service - ntopng - High-Speed Web-based Traffic Analysis and Flow Collection Tool
     Loaded: loaded (/lib/systemd/system/ntopng.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Mon 2022-08-29 14:50:25 UTC; 4s ago
       Docs: man:ntopng(8)
             file:/usr/share/doc/ntopng/README.Debian
             file:/usr/share/doc/ntopng/UserGuide.pdf.gz
    Process: 2337127 ExecStart=/usr/sbin/ntopng /etc/ntopng.conf (code=exited, status=1/FAILURE)
        CPU: 15ms

Aug 29 14:50:24 lb-kvm ntopng[2337127]: 29/Aug/2022 14:50:24 [Redis.cpp:119] ERROR: NOAUTH Authentication required.
Aug 29 14:50:25 lb-kvm ntopng[2337127]: [Redis.cpp:148] ERROR: ntopng requires redis server to be up and running
Aug 29 14:50:25 lb-kvm ntopng[2337127]: 29/Aug/2022 14:50:25 [Redis.cpp:148] ERROR: ntopng requires redis server to be up and running
Aug 29 14:50:25 lb-kvm ntopng[2337127]: 29/Aug/2022 14:50:25 [Redis.cpp:149] ERROR: Please start it and try again or use -r
Aug 29 14:50:25 lb-kvm ntopng[2337127]: 29/Aug/2022 14:50:25 [Redis.cpp:150] ERROR: to specify a redis server other than the default
Aug 29 14:50:25 lb-kvm systemd[1]: ntopng.service: Control process exited, code=exited, status=1/FAILURE
Aug 29 14:50:25 lb-kvm ntopng[2337127]: [Redis.cpp:149] ERROR: Please start it and try again or use -r
Aug 29 14:50:25 lb-kvm systemd[1]: ntopng.service: Failed with result 'exit-code'.
Aug 29 14:50:25 lb-kvm ntopng[2337127]: [Redis.cpp:150] ERROR: to specify a redis server other than the default
Aug 29 14:50:25 lb-kvm systemd[1]: Failed to start ntopng - High-Speed Web-based Traffic Analysis and Flow Collection Tool.
```

**Debug Information**:
redis
```
# netstat -tunpevaW | grep 6379
tcp        0      0 127.0.0.1:6379          0.0.0.0:*               LISTEN      933        7697079    2195287/redis-serve 
tcp6       0      0 ::1:6379                :::*                    LISTEN      933        7697080    2195287/redis-serve 

# redis-cli -a 'pZkLF+3GUsGLg1Qoz80qJsF2oiISLUuj66idsFn47e906nuEtODowVbMwEiWgTFhFvkYiFucnBkUD8FvCCmkSw==' PING
Warning: Using a password with '-a' or '-u' option on the command line interface may not be safe.
PONG
```

ntopng
```
grep -- ^-r /etc/ntopng/ntopng.conf
-r 127.0.0.1:6379:pZkLF+3GUsGLg1Qoz80qJsF2oiISLUuj66idsFn47e906nuEtODowVbMwEiWgTFhFvkYiFucnBkUD8FvCCmkSw==
```

**Comment by jean-christophe-manciot** [comment]
If I comment out the redis server ```requirepass```, remove the redis password in ntopng.conf and restart both services, ntopng starts without issue.

**Comment by jean-christophe-manciot** [comment]
I thought it was an issue related to the password length/strength, but it does not. Even with '123456789' for redis password, ntopng fails to start:
```
# systemctl restart redis
# redis-cli -a '123456789' PING
Warning: Using a password with '-a' or '-u' option on the command line interface may not be safe.
PONG
# grep -- ^-r /etc/ntopng/ntopng.conf
-r=127.0.0.1:6379:123456789

# systemctl restart ntopng
# systemctl status ntopng
● ntopng.service - ntopng - High-Speed Web-based Traffic Analysis and Flow Collection Tool
     Loaded: loaded (/lib/systemd/system/ntopng.service; enabled; vendor preset: enabled)
     Active: activating (start) since Mon 2022-08-29 16:23:09 UTC; 4s ago
       Docs: man:ntopng(8)
             file:/usr/share/doc/ntopng/README.Debian
             file:/usr/share/doc/ntopng/UserGuide.pdf.gz
Cntrl PID: 1632261 (ntopng)
      Tasks: 1 (limit: 9398)
     Memory: 3.2M
        CPU: 11ms
     CGroup: /system.slice/ntopng.service
             └─1632261 /usr/sbin/ntopng /etc/ntopng.conf

Aug 29 16:23:10 lb-kvm ntopng[1632261]: [Redis.cpp:119] ERROR: NOAUTH Authentication required.
Aug 29 16:23:11 lb-kvm ntopng[1632261]: 29/Aug/2022 16:23:11 [Redis.cpp:83] Redis has disconnected, reconnecting [remaining attempts: 13]
Aug 29 16:23:11 lb-kvm ntopng[1632261]: 29/Aug/2022 16:23:11 [Redis.cpp:119] ERROR: NOAUTH Authentication required.
Aug 29 16:23:11 lb-kvm ntopng[1632261]: [Redis.cpp:119] ERROR: NOAUTH Authentication required.
Aug 29 16:23:12 lb-kvm ntopng[1632261]: 29/Aug/2022 16:23:12 [Redis.cpp:83] Redis has disconnected, reconnecting [remaining attempts: 12]
Aug 29 16:23:12 lb-kvm ntopng[1632261]: 29/Aug/2022 16:23:12 [Redis.cpp:119] ERROR: NOAUTH Authentication required.
Aug 29 16:23:12 lb-kvm ntopng[1632261]: [Redis.cpp:119] ERROR: NOAUTH Authentication required.
Aug 29 16:23:13 lb-kvm ntopng[1632261]: 29/Aug/2022 16:23:13 [Redis.cpp:83] Redis has disconnected, reconnecting [remaining attempts: 11]
Aug 29 16:23:13 lb-kvm ntopng[1632261]: 29/Aug/2022 16:23:13 [Redis.cpp:119] ERROR: NOAUTH Authentication required.
Aug 29 16:23:13 lb-kvm ntopng[1632261]: [Redis.cpp:119] ERROR: NOAUTH Authentication required.
```

**Comment by jean-christophe-manciot** [comment]
Same issue with ntopng ```5.2.220414```.

**Comment by dandare100** [comment]
Hello @jean-christophe-manciot 

I tested it on my side and it works fine.

I did notice in your posts above that the service is being started with /etc/ntopng.conf for the config.

`ExecStart=/usr/sbin/ntopng /etc/ntopng.conf`

and the grep you are doing  to show us your password is grepping /etc/ntopng/ntopng.conf

`grep -- ^-r /etc/ntopng/ntopng.conf`

Please check if it works after you move your current config containing the password  (/etc/ntopng/ntopng.conf) to /etc/ntopng.conf  or changing /etc/ntopng.conf to /etc/ntopng/ntopng.conf in the systemd service file

Thanks

Jeff

**Comment by cardigliano** [comment]
Good catch @dandare100 ! 
@jean-christophe-manciot did you change the systemd script perhaps? Please fix/restore it to point to the right configuration file

**Comment by jean-christophe-manciot** [comment]
@cardigliano @dandare100 
OK,you're right, but using ```/etc/ntopng.conf``` and redis with a password a little bit more complex and stronger than 123456789, such as ```YAEGV6=[rItKb*[tbh}zX@XKVR^wbkrP``` makes ntopng fail:
```
# systemctl restart redis
# redis-cli -a 'YAEGV6=[rItKb*[tbh}zX@XKVR^wbkrP' PING
Warning: Using a password with '-a' or '-u' option on the command line interface may not be safe.
PONG
# grep -- ^-r /etc/ntopng.conf
-r=127.0.0.1:6379:YAEGV6=[rItKb*[tbh}zX@XKVR^wbkrP
# systemctl restart ntopng
# systemctl status ntopng
× ntopng.service - ntopng - High-Speed Web-based Traffic Analysis and Flow Collection Tool
     Loaded: loaded (/lib/systemd/system/ntopng.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Wed 2022-08-31 14:33:09 UTC; 26s ago
       Docs: man:ntopng(8)
             file:/usr/share/doc/ntopng/README.Debian
             file:/usr/share/doc/ntopng/UserGuide.pdf.gz
    Process: 919748 ExecStart=/usr/sbin/ntopng /etc/ntopng.conf (code=exited, status=1/FAILURE)
        CPU: 10ms

Aug 31 14:33:09 lb-kvm ntopng[919748]: 31/Aug/2022 14:33:09 [Redis.cpp:148] ERROR: ntopng requires redis server to be up and running
Aug 31 14:33:09 lb-kvm ntopng[919748]: 31/Aug/2022 14:33:09 [Redis.cpp:149] ERROR: Please start it and try again or use -r
Aug 31 14:33:09 lb-kvm ntopng[919748]: 31/Aug/2022 14:33:09 [Redis.cpp:150] ERROR: to specify a redis server other than the default
Aug 31 14:33:09 lb-kvm ntopng[919748]: [Redis.cpp:108] ERROR: Redis authentication failed: WRONGPASS invalid username-password pair
Aug 31 14:33:09 lb-kvm ntopng[919748]: [Redis.cpp:148] ERROR: ntopng requires redis server to be up and running
Aug 31 14:33:09 lb-kvm ntopng[919748]: [Redis.cpp:149] ERROR: Please start it and try again or use -r
Aug 31 14:33:09 lb-kvm ntopng[919748]: [Redis.cpp:150] ERROR: to specify a redis server other than the default
Aug 31 14:33:09 lb-kvm systemd[1]: ntopng.service: Control process exited, code=exited, status=1/FAILURE
Aug 31 14:33:09 lb-kvm systemd[1]: ntopng.service: Failed with result 'exit-code'.
Aug 31 14:33:09 lb-kvm systemd[1]: Failed to start ntopng - High-Speed Web-based Traffic Analysis and Flow Collection Tool.
```

I could not find anywhere in the official docs the passwords constraints:
- the characters which are allowed/not allowed
- their max length

Please, **allow all (or at least latin) characters otherwise the passwords strength can be significantly reduced**.
Since the password is always the last part of the config line, even ':' and ';' characters should be allowed despite the fact that they are used are delimiters before the password.

**Comment by jean-christophe-manciot** [comment]
Same issue with mysql 8.0.

This is a **major issue** because in order to use ntopng, we would need to weaken the security of all its connected parts:
- ElasticSearch
- redis
- mysql
- mysql-nprobe

**Comment by jean-christophe-manciot** [comment]
@lucaderi 
Any plan to allow all (at least Latin) characters to be used in all passwords to prevent weakened security of ntopng and all its connected parts?

**Comment by cardigliano** [comment]
@jean-christophe-manciot I pushed a fix for this in ac3d3cc38ff83aab9a86de53fcf2295e7249e058
ntopng was not supporting passwords containing @ characters as it is used to identify the DB ID, however I improved the parser and it should work now in your case. In any case I suggest using a password without : and @ characters.

**Comment by cardigliano** [comment]
A new package will be available later today (1h from now)
